sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -4,54 +4,244 @@
 #nullable enable
 
 using System;
+using System.Collections;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq.Expressions;
 using System.Reflection;
 
 namespace Azure.Core.Pipeline
 {
+#pragma warning disable CA1001 // Implement IDisposable
     internal readonly struct DiagnosticScope : IDisposable
+#pragma warning restore CA1001 // Implement IDisposable
     {
-        private readonly DiagnosticActivity? _activity;
+        private static readonly ConcurrentDictionary<string, object?> ActivitySources = new();
 
-        private readonly string _name;
+        private readonly ActivityAdapter? _activityAdapter;
 
-        private readonly DiagnosticListener _source;
+        internal DiagnosticScope(string ns, string scopeName, DiagnosticListener source, ActivityKind kind)
+        {
+            var activitySource = GetActivitySource(ns, scopeName);
+
+            IsEnabled = source.IsEnabled() || ActivityExtensions.ActivitySourceHasListeners(activitySource);
+
+            _activityAdapter = IsEnabled ? new ActivityAdapter(activitySource, source, scopeName, kind, null) : null;
+        }
 
-        internal DiagnosticScope(string name, DiagnosticListener source)
+        internal DiagnosticScope(string scopeName, DiagnosticListener source, object? diagnosticSourceArgs, object? activitySource, ActivityKind kind)
         {
-            _name = name;
-            _source = source;
-            _activity = _source.IsEnabled(_name) ? new DiagnosticActivity(_name) : null;
-            _activity?.SetW3CFormat();
+            IsEnabled = source.IsEnabled() || ActivityExtensions.ActivitySourceHasListeners(activitySource);
+
+            _activityAdapter = IsEnabled ? new ActivityAdapter(activitySource, source, scopeName, kind, diagnosticSourceArgs) : null;
         }
 
-        public bool IsEnabled => _activity != null;
+        public bool IsEnabled { get; }
+
+        private static object? GetActivitySource(string ns, string name)
+        {
+            if (!ActivityExtensions.SupportsActivitySource())
+            {
+                return null;
+            }
+
+            int indexOfDot = name.IndexOf(".", StringComparison.OrdinalIgnoreCase);
+            if (indexOfDot == -1)
+            {
+                return null;
+            }
+
+            string clientName = ns + "." + name.Substring(0, indexOfDot);
+
+            return ActivitySources.GetOrAdd(clientName, static n => ActivityExtensions.CreateActivitySource(n));
```

Comment:
So the ConcurrentDictionary should only grow as large as the client has methods? Seems reasonable if so.

Reply:
Even smaller - one entry per client type.